src/System.Private.ServiceModel/tests/Common/Scenarios/ScenarioTestTypes.cs

```diff
@@ -803,4 +831,26 @@ public void OnPingCallback(Guid guid)
             _tcs.SetResult(guid);
         });
     }
+
+    public void OnXmlPingCallback(XmlCompositeType xmlCompositeType)
+    {
+        // Set the result in an async task with a 100ms delay to prevent a race condition
+        // where the OnPingCallback hasn't sent the reply to the server before the channel is closed.
+        Task.Run(async () =>
+        {
+            await Task.Delay(100);
+            _xml_tcs.SetResult(xmlCompositeType);
```

Comment:
Not sure, I think either Matt or Jason added this functionality to the existing 'OnPingCallback' which I then adopted., I can confirm the race condition does occur.
I'm not sure how else we could do it, I can't see a way to make the cleanup code smarter for this particular scenario, but there may be a solution there.